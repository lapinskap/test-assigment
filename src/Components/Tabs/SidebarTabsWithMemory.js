/* eslint-disable function-call-argument-newline */
import PropTypes from 'prop-types';
import React, { useState, useEffect } from 'react';
import { MdModeEdit } from 'react-icons/md';
import { useHistory, Prompt, Link } from 'react-router-dom';
import {
  Card, CardHeader, CardBody,
} from 'reactstrap';
import Sticky from 'react-stickynode';
import RbsButton from '../../utils/RoleBasedSecurity/SecurityComponents/RbsButton';

export default function SidebarTabsWithMemory({
  tabsConfig, defaultKey, promptWhenUnsaved = true, collectiveSaveAction, additionalHeader, submitPermission,
}) {
  const history = useHistory();
  const locationHash = history.location.hash.replace('#', '');
  const [activeKey, setActiveKey] = useState(locationHash || defaultKey);
  const [editedTabs, setEditedTabs] = useState({});

  useEffect(() => () => {
    const { hash } = history.location;
    setActiveKey(hash ? hash.replace('#', '') : defaultKey);
  }, [history.location, defaultKey]);

  const currentKey = activeKey || history.location.hash.replace('#', '');

  const setTabIsEdited = (key, value) => {
    if (editedTabs[key] !== value) {
      const newState = { ...editedTabs };
      newState[key] = value;
      setEditedTabs(newState);
    }
  };

  const getLeaveWarningMessage = (unsavedTabsKeys, location) => {
    if (location.pathname === history.location.pathname) {
      return '';
    }
    const unsavedTabsName = unsavedTabsKeys.map((key) => {
      const tab = tabsConfig.find((el) => el.key === key);
      return tab ? tab.name : null;
    }).filter(Boolean);
    if (unsavedTabsName.length === 1) {
      return `Masz niezapisany formularz w zakładce ${unsavedTabsName[0]}. Niezapisane zmiany zostaną utracone.`;
    }
    return `Masz niezapisane formularze w zakładkach:${unsavedTabsName
      ? ` ${unsavedTabsName.join(', ')}.` : ','} Niezapisane zmiany zostaną utracone.`;
  };

  const unsavedForms = Object.keys(editedTabs).filter((key) => Boolean(editedTabs[key]));
  const currentTab = tabsConfig.find((item) => item.key === currentKey);
  return (
    <Card>
      <div className="forms-wizard-vertical">
        {collectiveSaveAction ? (
          <Sticky top=".app-header" innerZ="100">
            {additionalHeader || null}
            <CardHeader>
              <div className="btn-actions-pane-right">
                <RbsButton
                  permission={submitPermission}
                  size="lg"
                  color="success"
                  type="submit"
                  className="m-2"
                >
                  Zapisz
                </RbsButton>
              </div>
            </CardHeader>
          </Sticky>
        ) : null}
        <div style={{ display: 'flex' }}>
          <div style={{ width: '20%', display: 'flex', flexDirection: 'column' }}>
            <ol className="forms-wizard" style={{ width: '100%' }} data-t1="sidebarTabs">
              {
                    tabsConfig.map(({
                      key, name, icon, invalid,
                    }) => (
                      <Link key={key} to={`#${key}`} data-t1={key}>
                        <li
                          className={`form-wizard-step-${currentKey === key ? 'doing' : 'todo'}`}
                          key={key}
                        >
                          <em><i className={`${icon || 'pe-7s-tools'} btn-icon-wrapper`}> </i></em>
                          <span>{name}</span>
                          <span className="btn-actions-pane-right">
                            {editedTabs[key]
                              ? (
                                <MdModeEdit
                                  fontSize="20px"
                                />
                              ) : null}
                            {invalid ? (
                              <>
                                <string className="text-danger">!</string>
                              </>
                            ) : null}
                          </span>
                        </li>
                      </Link>
                    ))
                                }
            </ol>
          </div>
          <CardBody style={{ display: 'flex', flexDirection: 'column' }}>
            {currentTab ? React.cloneElement(
              currentTab.component,
              { setIsEdited: (value) => setTabIsEdited(activeKey, value) },
            ) : false}
          </CardBody>
        </div>
        {promptWhenUnsaved ? (
          <Prompt
            when={unsavedForms.length > 0}
            message={(location) => getLeaveWarningMessage(unsavedForms, location)}
          />
        ) : null}
      </div>
    </Card>
  );
}

SidebarTabsWithMemory.propTypes = {
  collectiveSaveAction: PropTypes.func,
  submitPermission: PropTypes.string,
  defaultKey: PropTypes.string.isRequired,
  promptWhenUnsaved: PropTypes.bool,
  additionalHeader: PropTypes.node,
  tabsConfig: PropTypes.arrayOf(PropTypes.shape({
    key: PropTypes.string,
    name: PropTypes.string,
    icon: PropTypes.string,
    invalid: PropTypes.bool,
  })).isRequired,
};

SidebarTabsWithMemory.defaultProps = {
  collectiveSaveAction: null,
  submitPermission: null,
  additionalHeader: null,
  promptWhenUnsaved: true,
};
