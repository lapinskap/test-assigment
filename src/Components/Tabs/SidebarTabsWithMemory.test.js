/* eslint-disable function-call-argument-newline */
/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable react/prop-types */
import React from 'react';
import renderer from 'react-test-renderer';
import SidebarTabsWithMemory from './SidebarTabsWithMemory';
import '../../tests/setupTests';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    location: {
      hash: '#test',
    },
  }),
  Prompt: ({ message }) => <div>{message}</div>,
  Link: ({ to, children, ...props }) => <a href={to} {...props}>{children}</a>,
}));

describe('<SidebarTabsWithMemory />', () => {
  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, 'useState');
  useStateSpy.mockImplementation((init) => [init, setState]);

  function createNodeMock() {
    return {
      getBoundingClientRect() {
        return {
          width: 100,
        };
      },
    };
  }

  let tabsConfig; let defaultKey; const promptWhenUnsaved = true; let
    collectiveSaveAction;

  beforeEach(() => {
    tabsConfig = [
      {
        key: '20',
        name: 'test',
        icon: 's-glyph-5',
      },
    ];
    defaultKey = '';
    collectiveSaveAction = () => ({});
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly with default state', () => {
    const tree = renderer
      .create(<SidebarTabsWithMemory
        tabsConfig={tabsConfig}
        defaultKey={defaultKey}
        promptWhenUnsaved={promptWhenUnsaved}
        collectiveSaveAction={collectiveSaveAction}
      />, { createNodeMock })
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
