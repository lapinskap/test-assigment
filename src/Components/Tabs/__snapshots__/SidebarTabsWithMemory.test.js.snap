// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SidebarTabsWithMemory /> renders correctly with default state 1`] = `
<div
  className="card"
>
  <div
    className="forms-wizard-vertical"
  >
    <div
      className="sticky-outer-wrapper"
      style={Object {}}
    >
      <div
        className="sticky-inner-wrapper"
        style={
          Object {
            "position": "relative",
            "top": "",
            "transform": "translate3d(0,0px,0)",
            "zIndex": "100",
          }
        }
      >
        <div
          className="card-header"
        >
          <div
            className="btn-actions-pane-right"
          >
            <button
              aria-label={null}
              className="m-2 btn btn-success btn-lg"
              disabled={false}
              onClick={[Function]}
              title={null}
              type="submit"
            >
              Zapisz
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <div
        style={
          Object {
            "display": "flex",
            "flexDirection": "column",
            "width": "20%",
          }
        }
      >
        <ol
          className="forms-wizard"
          data-t1="sidebarTabs"
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <a
            data-t1="20"
            href="#20"
          >
            <li
              className="form-wizard-step-todo"
            >
              <em>
                <i
                  className="s-glyph-5 btn-icon-wrapper"
                >
                   
                </i>
              </em>
              <span>
                test
              </span>
              <span
                className="btn-actions-pane-right"
              />
            </li>
          </a>
        </ol>
      </div>
      <div
        className="card-body"
        style={
          Object {
            "display": "flex",
            "flexDirection": "column",
          }
        }
      />
    </div>
    <div />
  </div>
</div>
`;
